### PR TITLE
feat(`x/marketmap`): init types and validation

### PR DESCRIPTION
**Pull request by aljo242**
Closes BLO-1029

### CODE REVIEW

File: proto/slinky/mm2/v1/market.proto
Comment (davidterpay) [inline proto/slinky/mm2/v1/market.proto:83]: Do we still need to have this be a map? i'd prefer moving this into a list for simplicity. All off-chain access to this will have to have its own transformations anyways so map doesn't introduce much ease.

Comment (aljo242) [reply proto/slinky/mm2/v1/market.proto:83]: Yeah we can definitely do this

File: x/mm2/types/market.go
Comment (davidterpay) [inline x/mm2/types/market.go]:
```suggestion
// ValidateBasic performs aggregate validation for all fields in the MarketMap. We consider
// the market map to be valid iff:
//
// 1. Each ticker is valid.
// 2. Each provider is valid.
// 3. Any provider configuration that requires a transformation (normalizeByPair) must be included in the main set of tickers.
```

File: x/mm2/types/market.go
Comment (davidterpay) [inline x/mm2/types/market.go]:
```suggestion
					return fmt.Errorf("provider's (%s) pair for normalization (%s) was not found in the marketmap", providerConfig.Name, providerConfig.NormalizeByPair.String())
```

Comment (davidterpay) [reply x/mm2/types/market.go]: or something like this, it can be hard pinpointing what provider was misconfigured otherwise.

File: proto/slinky/mm2/v1/market.proto
Comment (g3ndo9kar9) [inline proto/slinky/mm2/v1/market.proto:48]: This is where we'd be storing the `MinPriceChangePPM`?

File: proto/slinky/mm2/v1/market.proto
Comment (g3ndo9kar9) [inline proto/slinky/mm2/v1/market.proto]: nit: `normalize_by_pair`

File: proto/slinky/mm2/v1/market.proto
Comment (g3ndo9kar9) [inline proto/slinky/mm2/v1/market.proto:73]: nit: 5

File: proto/slinky/mm2/v1/market.proto
Comment (g3ndo9kar9) [inline proto/slinky/mm2/v1/market.proto:44]: Where are these proto field IDs coming from?

File: proto/slinky/mm2/v1/market.proto
Comment (g3ndo9kar9) [inline proto/slinky/mm2/v1/market.proto:36]: How do these compare to the dydx decimals? Those can be negative, no?

File: proto/slinky/mm2/v1/market.proto
Comment (g3ndo9kar9) [reply proto/slinky/mm2/v1/market.proto:83]: We're never indexing directly into the market-config? If so, moving to a slice makes the index operation O(n)

File: x/mm2/types/market.go
Comment (g3ndo9kar9) [inline x/mm2/types/market.go]: nit: is this name still relevant?

File: x/mm2/types/market.go
Comment (g3ndo9kar9) [inline x/mm2/types/market.go:111]: are there tests for this file?

File: proto/slinky/mm2/v1/market.proto
Comment (aljo242) [reply proto/slinky/mm2/v1/market.proto:48]: yes

File: proto/slinky/mm2/v1/market.proto
Comment (aljo242) [reply proto/slinky/mm2/v1/market.proto:44]: I like putting metadata or flags at the "end" of the proto def. 

Field IDs 1-15 take the same amount of data to encode, so there really isnt any different.  It is more of a stylistic thing.  If we wanted to extend fields in the ticker, we could just 4 in the future which is slightly cuter lol

File: proto/slinky/mm2/v1/market.proto
Comment (aljo242) [reply proto/slinky/mm2/v1/market.proto:83]: During validation, you need to construct a map to check everything, then when the struct is consumed by users, they will use it in  a map like fashion.  It is easier to leave as a map.

File: x/mm2/types/market.go
Comment (aljo242) [reply x/mm2/types/market.go]: yes - we are validating the market map for the index price aggregation strategy it must be set up to support

File: proto/slinky/mm2/v1/market.proto
Comment (davidterpay) [reply proto/slinky/mm2/v1/market.proto:36]: we just take the absolute value in our case.

File: x/mm2/types/market.go
Comment (aljo242) [reply x/mm2/types/market.go:111]: Adding

**Review by davidterpay — APPROVED** [review]
lgtm after some testing is pushed

File: x/mm2/types/market.go
Comment (aljo242) [reply x/mm2/types/market.go]: removed and just using `ValidateBasic()`

File: proto/slinky/mm2/v1/market.proto
Comment (Eric-Warehime) [reply proto/slinky/mm2/v1/market.proto:48]: 🤔 Not sure if we have anything written already utilizing this JSON/how the ser/de struct will be integrated into the sidecar that I could look at? 

I want to minimize the number of places we have to keep stateful logic around which chain we're running for, and this seems like a field that would potentially be different not on a per-provider basis but on a per-chain basis.

`MinPriceChangePPM` is a good example of that which is at the moment dydx-specific.

File: x/mm2/types/ticker.go
Comment (Eric-Warehime) [inline x/mm2/types/ticker.go:55]: Not covered by tests.

File: x/mm2/types/market.go
Comment (Eric-Warehime) [inline x/mm2/types/market.go]: The concept of a "path" no longer exists right?

File: x/mm2/types/market.go
Comment (Eric-Warehime) [inline x/mm2/types/market.go]: Is the `Operation` type gone now too?

File: x/mm2/types/market.go
Comment (Eric-Warehime) [inline x/mm2/types/market.go:23]: Needs test.

File: x/mm2/types/market.go
Comment (Eric-Warehime) [inline x/mm2/types/market.go:61]: Needs test.

File: x/mm2/types/market.go
Comment (aljo242) [reply x/mm2/types/market.go:23]: tested by "provider config includes a ticker that is not supported"

File: x/mm2/types/ticker.go
Comment (aljo242) [reply x/mm2/types/ticker.go:55]: tested by "invalid json len"

**Review by Eric-Warehime — APPROVED** [review]

File: proto/slinky/mm2/v1/market.proto
Comment (Christopher-Li) [inline proto/slinky/mm2/v1/market.proto:48]: is there a reason why you skipped 4 - 13?

Also do you have any ideas or examples of what would be stored in here?

File: proto/slinky/mm2/v1/market.proto
Comment (Christopher-Li) [inline proto/slinky/mm2/v1/market.proto:73]: skipped 5 - 14?

File: proto/slinky/mm2/v1/market.proto
Comment (Christopher-Li) [inline proto/slinky/mm2/v1/market.proto:62]: nit: typo `BTC/USD`

File: x/mm2/types/provider.go
Comment (Christopher-Li) [inline x/mm2/types/provider.go:32]: should we add any validation that ProviderConfig.metadata_JSON holds any particular data?

File: x/mm2/types/ticker.go
Comment (Christopher-Li) [inline x/mm2/types/ticker.go:18]: nit: what unit is this? In bytes or bits?

File: proto/slinky/mm2/v1/market.proto
Comment (aljo242) [reply proto/slinky/mm2/v1/market.proto:48]: There is no impact to the size of the data structure when choosing field IDs 1 - 16 (all take 2bz to encode).

This is just a stylistic move of putting metadata (an optional field) at the "end" of the struct

File: x/mm2/types/provider.go
Comment (aljo242) [reply x/mm2/types/provider.go:32]: We can at a later time when we have solid metadata types we care about.  Now that we are adding defi providers like raydium, they use this field, so we can add those once merged.  Good callout.  adding a linear issue for us to track this

File: x/mm2/types/ticker.go
Comment (aljo242) [reply x/mm2/types/ticker.go:18]: bytes !

Comment (aljo242) [reply x/mm2/types/ticker.go:18]: updating comment !

**Pull request merged**
